File: part-2-main/blur_int.c
//-------------------------------------------------------------------------------------------------
// 4190.308 Computer Architecture                                                       Spring 2023
//
/// @file
/// @brief Image blurring (int)
///        This module implements a function that blurs an image with a filter (integer version)
///
/// @author Your Name <your@email.com>
///
/// @section changelog Change Log
/// 2023/MM/DD Your Name Comment
///
//-------------------------------------------------------------------------------------------------

#include <stdlib.h>
#include "blur.h"


struct Image blur_int(struct Image image, int kernel_size)
{
    int h=image.height - kernel_size + 1;
    int w=image.width - kernel_size + 1;
    int channel=image.channels;
    struct Image output;
    output.height=h;
    output.width=w;
    output.channels=channel;
    output.data = (uint8*)calloc(h*w*channel, sizeof(uint8));
    if (output.data == NULL) abort();
    int mid=(kernel_size/2);

    int fpnum1;
    int fpnum2;
    if(kernel_size==3){
        fpnum1 = 28;
        fpnum2=3;}
    else if(kernel_size==5){
        fpnum1 = 10;
        fpnum2=5;}
    else if(kernel_size==7){
        fpnum1 = 5;
        fpnum2=10;}
    for (int y = 0; y < h; y++) {
        for (int x = 0; x < w; x++) {
            for (int c = 0; c < channel; c++) {
                int temp=0;
                for (int k = 0; k < kernel_size; k++) {
                    for (int k2 = 0; k2 < kernel_size; k2++) {
                        temp += fpnum1*PIXEL(image, y+k,x+k2,c);
//                        printf("%d %d %d : %d \n",y+k,x+k2,c,temp);
                    }
                }
                temp += fpnum2*PIXEL(image, y+mid,x+mid,c);
//                printf("mid : %d %d\n",y+mid,x+mid);
//                printf("%d %d %d : %d \n",y,x,c,(uint8) (temp>> 8));
                PIXEL(output, y, x, c) = (uint8) (temp>> 8);

            }
        }
    }


  return output;

}

File: part-1-main/blur_int.py
"""
4190.308 Computer Architecture                                                          Spring 2023

Image blurring (int)

This module implements a function that blurs an image with a filter (integer version)

@author:
    Your Name <your@email.com>

@changes:
    2023/MM/DD Your Name Comment

"""



def blur(image, height, width, channels, kernel_size=5):
    """
    Blurs an image with a kernel and returns the blurred image.

    Args:
        image:        image data (multi-level list)
        height:       image height
        width:        image width
        channels:     number of channels (BGR or BGRA)
        kernel_size:  size of blurring kernel

    Returns:
        A tuple containing the following elements:
        - blurred:    blurred image data
        - bheight:    blurred image height
        - bwidth:     blurred image width
        - bchannels:  blurred image channels

    """

    # TODO
    # Your work goes here

    # For now, we simply copy the input parameters into the output parameters.
    # Fix/adjust once you have implemented your solution.
    bheight = height - kernel_size + 1
    bwidth = width - kernel_size + 1
    bchannels = channels
    blurred=[[[0 for col in range(channels)] for row in range(bwidth)] for depth in range(bheight)]

    mid=int(kernel_size/2)
    if (kernel_size == 3):
        fpnum1 = 28
        fpnum2=3
    elif (kernel_size == 5):
        fpnum1 = 10
        fpnum2 = 5
    elif (kernel_size == 7):
        fpnum1 = 5
        fpnum2 = 10

    for y in range(bheight):
        for x in range(bwidth):
            for c in range(channels):
                for k in range(kernel_size):
                    for k2 in range(kernel_size):
                        blurred[y][x][c] += (fpnum1 * (image[y + k][x + k2][c]) )
                blurred[y][x][c] += image[y + mid][x + mid][c] * fpnum2
                blurred[y][x][c] = blurred[y][x][c] >> 8
                # print(f'blurred[{y}][{x}][{c}] : {blurred[y][x][c]}')

    return blurred, bheight, bwidth, bchannels



